### PR TITLE
[Datetime] Added minTime and maxTime options to TimePicker component

### LINKED ISSUE #1220: [TimePicker] Add minTime and maxTime options to TimePicker
I think that minTime and maxTime options for TimePicker (as minDate and maxDate for DateInput) will be very useful.
I'm ready to contribute this feature

### PR DESCRIPTION
**Pull request by JacekJagiello**
#### Checklist

- [x] [Enable CircleCI for your fork](https://circleci.com/add-projects)
- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

Added minTime and maxTime options

#### Reviewers should focus on:

Please make sure feature works as it's expected
Please read my questions in #1220
Check if API is designed correctly(usage of Date object)
Code quality
Check if tests names make sense to you

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/blueprint](https://github.com/palantir/blueprint), @JacekJagiello! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: packages/datetime/test/timePickerTests.tsx [hunk 1]
```diff
@@ -209,6 +209,214 @@ describe("<TimePicker>", () => {
...
+            assert.strictEqual(minuteInput.value, "59");
+
+            changeInputThenBlur(secondInput, "59");
+            assert.strictEqual(secondInput.value, "59");
+
+            changeInputThenBlur(millisecondInput, "999");
+            assert.strictEqual(millisecondInput.value, "999");
+        });
+
+        it("I can select time even time range past midnight", () => {
```
Comment (JacekJagiello) [inline packages/datetime/test/timePickerTests.tsx]: Is the name of this test make sense to you?

File: packages/datetime/src/timePicker.tsx [hunk 2]
```diff
@@ -252,10 +274,30 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
...
+    private getDateOnlyWithTime(date: Date): Date {
+        return new Date(0, 0, 0, date.getHours(), date.getMinutes(), date.getSeconds(), date.getMilliseconds());
+    }
+
     private updateTime(time: number, unit: TimeUnit) {
-        if (isTimeValid(time, unit)) {
-            const newValue = DateUtils.clone(this.state.value);
-            setTimeUnit(time, newValue, unit);
+        const newValue = DateUtils.clone(this.state.value);
+        setTimeUnit(time, newValue, unit);
```
Comment (JacekJagiello) [inline packages/datetime/src/timePicker.tsx]: Do you think it's safe to call `setTimUnit()` before validating input values with `isTimeValid()`?

File: packages/datetime/test/timePickerTests.tsx [hunk 3]
```diff
@@ -209,6 +209,214 @@ describe("<TimePicker>", () => {
...
+
+            changeInputThenBlur(millisecondInput, "500");
+            assert.strictEqual(millisecondInput.value, "500");
+
+            /* I can select default maxTime */
+            changeInputThenBlur(hourInput, "23");
+            assert.strictEqual(hourInput.value, "23");
+
+            changeInputThenBlur(minuteInput, "59");
+            assert.strictEqual(minuteInput.value, "59");
```
Comment (giladgray) [inline packages/datetime/test/timePickerTests.tsx]: these change-assert pairs aren't helpful since the code inside `changeInputThenBlur` literally sets `input.value`, then you assert that that property has been changed. 
a) of course it has, you just changed it. 
b) DOM state is not the important thing to test here--we want to test the React component (its state and props).

would be much better to do all the DOM event manipulation, then test that the React component updated as expected: its internal state and event callbacks reflect what you expect.

File: packages/datetime/test/timePickerTests.tsx [hunk 4]
```diff
@@ -209,6 +209,214 @@ describe("<TimePicker>", () => {
...
+                precision: TimePickerPrecision.MILLISECOND,
+            });
+
+            const hourInput = findInputElement(Classes.TIMEPICKER_HOUR);
+
+            assert.strictEqual(hourInput.value, "15");
+            assert.strictEqual(timePicker.state.value.getHours(), 15);
+        });
+
+        it("if defaultValue is is greater than maxTime, minTime becomes initial time", () => {
```
Comment (giladgray) [inline packages/datetime/test/timePickerTests.tsx]: this is not what i'd expect. i'd expect to clamp to nearest bound, so default greater than max should result in max time as initial time.

File: packages/datetime/src/timePicker.tsx (hunk 2 above)
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: probably not... why was this change necessary?

File: packages/datetime/src/timePicker.tsx [hunk 5]
```diff
@@ -252,10 +274,30 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
         this.updateTime(loopTime(newTime, unit), unit);
     }
 
+    private isTimeInRange(date: Date) {
+        const format = "H:mm:ss:SSS";
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: why does format matter here? this is just for stringifying but we should be able to compare the actual data itself, not a string representation thereof.

File: packages/datetime/src/timePicker.tsx [hunk 6]
```diff
@@ -252,10 +274,30 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
...
+        const isTimeBeforeMaxTime = time.isSameOrBefore(maxTime);
+
+        if (maxTime.isBefore(minTime)) {
+            return isTimeAfterMinTime || isTimeBeforeMaxTime;
+        }
+
+        return isTimeAfterMinTime && isTimeBeforeMaxTime;
+    }
+
+    private getDateOnlyWithTime(date: Date): Date {
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: i think we have a Util function for this? or, we should; this doesn't belong as instance method. @cmslewis ?

File: packages/datetime/src/timePicker.tsx [hunk 7]
```diff
@@ -236,6 +251,13 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
         /* tslint:enable:object-literal-sort-keys */
     }
 
+    private adjustTimeToRange(date: Date): Date {
+        if (this.isTimeInRange(date)) {
+            return date;
+        }
+        return this.props.minTime;
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: choose closest bound instead of always using min time.

File: packages/datetime/src/timePicker.tsx [hunk 8]
```diff
@@ -236,6 +251,13 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
         /* tslint:enable:object-literal-sort-keys */
     }
 
+    private adjustTimeToRange(date: Date): Date {
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: `private getValidTimeInRange`

File: packages/datetime/test/timePickerTests.tsx [hunk 9]
```diff
@@ -209,6 +209,214 @@ describe("<TimePicker>", () => {
...
+            changeInputThenBlur(minuteInput, "33");
+            assert.strictEqual(minuteInput.value, "32");
+            assert.strictEqual(timePicker.state.value.getMinutes(), 32);
+
+            changeInputThenBlur(secondInput, "21");
+            assert.strictEqual(secondInput.value, "20");
+            assert.strictEqual(timePicker.state.value.getSeconds(), 20);
+
+            changeInputThenBlur(millisecondsInput, "700");
+            assert.strictEqual(millisecondsInput.value, "600");
```
Comment (giladgray) [inline packages/datetime/test/timePickerTests.tsx]: ❌ _please_ don't test DOM element value like this--it's not the canonical source of truth in React. 

**copy the other tests in this file.** they all look at `state.value` and use a test helper function that you should be able to easily reuse in all your tests:

```
assertTimeIs(timePicker.state.value, 0, 0, 0, 0);
```

File: packages/datetime/src/timePicker.tsx [hunk 10]
```diff
@@ -55,6 +56,18 @@ export interface ITimePickerProps extends IProps {
...
 
+    /**
+     * The latest time the user can select.
+     * Year, month and day of the Date object is ignored
+     */
+    maxTime?: Date;
+
+    /**
+     * The earliest time the user can select.
+     * Year, month and day of the Date object is ignored
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: please add oxford commas in both these comments: "Year, month[,] and day"

File: packages/datetime/src/timePicker.tsx [hunk 11]
```diff
@@ -252,10 +274,30 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
...
     }
 
+    private isTimeInRange(date: Date) {
+        const format = "H:mm:ss:SSS";
+        const time = moment(this.getDateOnlyWithTime(date), format);
+        const minTime = moment(this.getDateOnlyWithTime(this.props.minTime), format);
+        const maxTime = moment(this.getDateOnlyWithTime(this.props.maxTime), format);
+
+        const isTimeAfterMinTime = time.isSameOrAfter(minTime);
+        const isTimeBeforeMaxTime = time.isSameOrBefore(maxTime);
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: is it possible to do this _without_ relying on `moment`? just using the regular `DateUtils` we already have?

File: packages/datetime/src/common/dateUtils.ts [hunk 12]
```diff
@@ -110,6 +110,27 @@ export function isMonthInRange(date: Date, dateRange: DateRange) {
     return start <= day && day <= end;
 }
 
+export function isTimeInRange(time: Date, minTime: Date, maxTime: Date): boolean {
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: call the first arg `date` since it doesn't need to be time-only (and frees up the name `time` for comment below).

File: packages/datetime/src/common/dateUtils.ts [hunk 13]
```diff
@@ -110,6 +110,27 @@ export function isMonthInRange(date: Date, dateRange: DateRange) {
...
+    time = getDateOnlyWithTime(time);
+    minTime = getDateOnlyWithTime(minTime);
+    maxTime = getDateOnlyWithTime(maxTime);
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: :x: _never reassign arguments;_ they should be treated as constants. modifying them is a serious side effect and can affect calling code. define new variables here.

Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: also include the `.getTime()` calls in your new variables to avoid repeating it each time below.

File: packages/datetime/src/timePicker.tsx [hunk 14]
```diff
@@ -251,11 +250,16 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
         /* tslint:enable:object-literal-sort-keys */
     }
 
-    private adjustTimeToRange(date: Date): Date {
-        if (this.isTimeInRange(date)) {
-            return date;
+    private getValidTimeInRange(time: Date): Date {
+        const { minTime, maxTime } = this.props;
```
Comment (giladgray) [inline packages/datetime/src/timePicker.tsx]: 👍

File: packages/datetime/test/common/dateUtilsTests.tsx [hunk 15]
```diff
@@ -39,4 +39,55 @@ describe("dateUtils", () => {
...
     });
+
+    it("getDateOnlyWithTime returns Date object with constant year, month, and day", () => {
+        const date = new Date(1995, 6, 30, 14, 10, 10, 600);
+
+        const time = DateUtils.getDateOnlyWithTime(date);
+
+        assert.equal(time.getFullYear(), 1899);
+        assert.equal(time.getMonth(), 11);
+        assert.equal(time.getDay(), 0);
```
Comment (giladgray) [inline packages/datetime/test/common/dateUtilsTests.tsx]: lol awesome

File: packages/datetime/test/timePickerTests.tsx [hunk 16]
```diff
@@ -400,20 +351,11 @@ describe("<TimePicker>", () => {
...
-
             changeInputThenBlur(secondInput, "21");
-            assert.strictEqual(secondInput.value, "20");
-            assert.strictEqual(timePicker.state.value.getSeconds(), 20);
-
             changeInputThenBlur(millisecondsInput, "700");
-            assert.strictEqual(millisecondsInput.value, "600");
-            assert.strictEqual(timePicker.state.value.getMilliseconds(), 600);
+
+            assertTimeIs(timePicker.state.value, 15, 32, 20, 600);
```
Comment (giladgray) [inline packages/datetime/test/timePickerTests.tsx]: sweet! so much shorter! removed like 60 unnecessary lines 👍

File: packages/datetime/src/common/dateUtils.ts [hunk 17]
```diff
@@ -110,6 +110,27 @@ export function isMonthInRange(date: Date, dateRange: DateRange) {
...
+
+    const isTimeAfterMinTime = time.getTime() >= minTime.getTime();
+    const isTimeBeforeMaxTime = time.getTime() <= maxTime.getTime();
+
+    if (maxTime.getTime() <= minTime.getTime()) {
```
Comment (giladgray) [inline packages/datetime/src/common/dateUtils.ts]: hmm this sounds like an error state and should probably be validated by the component... 

likely safe to assume min <= max inside this function, though doesn't hurt to handle it like you're doing here.

Comment (JacekJagiello) [inline packages/datetime/src/common/dateUtils.ts]: Well, it's not state error, it's a special case when we have time range that past midnight. For example `minTime=22:00` to `maxTime=2:00` then `minTime` > `maxTime`. 

If I select time after midnight e.g. `1:00` then `isTimeAfterMinTime` becomes `false` and `isTimeBeforeMaxTime` becomes `true`.

If I select time before midnight e.g. `23:00` then `isTimeAfterMinTime` becomes `true` and `isTimeBeforeMaxTime` becomes `false`.

So if selected time is in a range that past midnight, one of this must be `true`.

I think variable names are a bit misleading. I think I should name them `isTimeGreaterThanMinTime` and `isTimeSmallerThanMaxTime` are more accurate? What do you think? :)

File: packages/datetime/test/common/dateUtilsTests.tsx [hunk 18]
```diff
@@ -39,4 +39,75 @@ describe("dateUtils", () => {
...
+            assert.isFalse(DateUtils.isTimeInRange(time, minTimeBeforeMidnight, maxTimeAfterMidnight));
+        });
+
+        it("returns false if given time is smaller than minTime", () => {
+            const time = new Date(1995, 6, 30, 13, 10, 50, 900);
+            assert.isFalse(DateUtils.isTimeInRange(time, minTime, maxTime));
+        });
+
+        it("returns false if given time is greater than maxTime", () => {
+            const time = new Date(1995, 6, 30, 18, 41, 9, 50);
```
Comment (cmslewis) [inline packages/datetime/test/common/dateUtilsTests.tsx]: Can we create a helper function to clarify that the day/month/year are ignored in all of these?

```tsx
function createTimeObject(hour: number, minute: number, second: number, millisecond: number) {
    const IGNORED_YEAR = 1995;
    const IGNORED_MONTH = 6;
    const IGNORED_DAY = 30;
    return new Date(IGNORED_YEAR, IGNORED_MONTH, IGNORED_DAY, hour, minute, second, millisecond);
}
```

Then, in each test, we can just do:

```tsx
const time = createTimeObject(13, 10, 50, 900);
```

File: packages/datetime/test/timePickerTests.tsx [hunk 19]
```diff
@@ -209,6 +209,156 @@ describe("<TimePicker>", () => {
...
+                defaultValue: new Date(1995, 6, 30, 20, 30),
+                maxTime: new Date(1995, 6, 30, 18, 30),
+                minTime: new Date(1995, 6, 30, 15, 30),
+                precision: TimePickerPrecision.MILLISECOND,
+            });
+
+            assertTimeIs(timePicker.state.value, 18, 30, 0, 0);
+        });
+
+        it("by default, I can select any time", () => {
```
Comment (cmslewis) [inline packages/datetime/test/timePickerTests.tsx]: Per @giladgray's earlier comment, first-person test descriptions aren't standard for our codebase, but we can either change these before merging this PR or after.

File: packages/datetime/src/timePicker.tsx [hunk 20]
```diff
@@ -266,7 +294,7 @@ export class TimePicker extends React.Component<ITimePickerProps, ITimePickerSta
     private updateState(state: ITimePickerState) {
         let newState = state;
         const hasNewValue = newState.value != null
-            && !DateUtils.areSameTime(newState.value, this.state.value);
+          && !DateUtils.areSameTime(newState.value, this.state.value);
```
Comment (AlexMarvelo) [inline packages/datetime/src/timePicker.tsx]: Lines are the same. Something wrong with indentation

File: packages/datetime/test/common/dateUtilsTests.tsx (hunk 18 above)
Comment (JacekJagiello) [inline packages/datetime/test/common/dateUtilsTests.tsx]: Sure thing, good idea

File: packages/datetime/test/common/dateUtilsTests.tsx [hunk 21]
```diff
@@ -39,4 +39,75 @@ describe("dateUtils", () => {
...
+        assert.equal(time.getMonth(), 11);
+        assert.equal(time.getDay(), 0);
+
+        assert.equal(time.getHours(), 14);
+        assert.equal(time.getMinutes(), 10);
+        assert.equal(time.getSeconds(), 10);
+        assert.equal(time.getMilliseconds(), 600);
+    });
+
+    describe("isTimeGreaterThan", () => {
```
Comment (JacekJagiello) [inline packages/datetime/test/common/dateUtilsTests.tsx]: Yeah, I've missed this case

File: packages/datetime/src/common/dateUtils.ts [hunk 22]
```diff
@@ -110,6 +110,27 @@ export function isMonthInRange(date: Date, dateRange: DateRange) {
...
+    const isTimeSmallerThanMaxTime = time <= maxTime;
+
+    if (maxTime <= minTime) {
+        return isTimeGreaterThanMinTime || isTimeSmallerThanMaxTime;
+    }
+
+    return isTimeGreaterThanMinTime && isTimeSmallerThanMaxTime;
+}
+
+export function isTimeGreaterThan(date1: Date, date2: Date): boolean {
```
Comment (JacekJagiello) [inline packages/datetime/src/common/dateUtils.ts]: Agree that `isTimeSameOrAfter` is better name, I'll change it 👍

File: packages/datetime/src/common/dateUtils.ts [hunk 23]
```diff
@@ -110,6 +110,27 @@ export function isMonthInRange(date: Date, dateRange: DateRange) {
...
 
+export function isTimeInRange(date: Date, minDate: Date, maxDate: Date): boolean {
+    const time = getDateOnlyWithTime(date).getTime();
+    const minTime = getDateOnlyWithTime(minDate).getTime();
+    const maxTime = getDateOnlyWithTime(maxDate).getTime();
+
+    const isTimeGreaterThanMinTime = time >= minTime;
+    const isTimeSmallerThanMaxTime = time <= maxTime;
+
+    if (maxTime <= minTime) {
```
Comment (JacekJagiello) [inline packages/datetime/src/common/dateUtils.ts]: Well, to be honest, I feel different about this. If `isTimeInRange` is supposed to tell me if given time is between another two times, and we consider `19:00 - 02:00` as a valid time range, then function works as expected. But I think I get your point - `maxTime > mintTime` from the logical point of view, should never happen, and that indeed is exceptional. That's why I won't argue, returning error is also fine for me :)

So, to clamp implementation details - I should catch error from `isTimeInRange` in the component, then if `allowOverlappingTimes` prop is set to `true`, set a new time, otherwise don't set a new time. Do I get it right?

Ps. Overlapping times - that's how this case should be named, not some kind of "past midnight" thing that I used to call it in tests :P I'll also fix that.

Comment (cmslewis) [inline packages/datetime/src/common/dateUtils.ts]: @giladgray can you offer your thoughts here if possible?

@JacekJagiello I definitely see where you're coming from. I've thought about this more: if we add a function-description comment above describing the handling of the overlapping case, I'd be fine going with the approach you've already implemented here.

Comment (JacekJagiello) [inline packages/datetime/src/common/dateUtils.ts]: @cmslewis I'm fine with throwing error solution, I can implement it this way. However, I'm also curious of @giladgray thoughts

**Comment by JacekJagiello** [comment]
I've refactored first person test descriptions. Check them out, and if you have better descriptions, please came up with them in comments :)

**Comment by llorca** [comment]
Preview of latest build: https://14-95210685-gh.circle-artifacts.com/0/home/ubuntu/blueprint/docs/docs/index.html

**Comment by JacekJagiello** [comment]
Here is updated example: https://15-95210685-gh.circle-artifacts.com/0/home/ubuntu/blueprint/packages/site-docs/dist/index.html#datetime/timepicker

**Comment by cmslewis** [comment]
Thanks for updating the `TimePicker` example, @JacekJagiello! Super helpful. Now that I've played with the feature, I've noticed some behavioral oddities that we should address before merging. Handling these oddities now will keep behaviors consistent across components. Please don't be discouraged—I'm looking forward to getting this merged, and we're almost there!

1. __Updating `minTime` and `maxTime`:__
    - _Currently:_ if you change the `minTime` or `maxTime` prop, the value does not immediately change.
min-max-time-bug-1

    - _Expected:_ When `minTime` or `maxTime` changes, update the `value` from within `componentWillReceiveProps`. (See `NumericInput#componentWillReceiveProps` for similar behavior.)

1. __Handling identical `minTime` and `maxTime`:__
    - _Currently:_ Having `minTime === maxTime` is equivalent to not having either specified at all; any time is valid in this case.
min-max-time-bug-2

    - _Expected:_ If `minTime === maxTime`, then the only valid `value` should be the very value to which both boundaries are set. (This is how `NumericInput` works, so we'll definitely want consistency.)

1. __Incrementing when `minTime > maxTime`:__
    - _Currently:_ If `minTime > maxTime` and `value === minTime`, then pressing the <kbd>Down Arrow</kbd> will jump down to the `maxValue`. However, if `value === maxTime`, then pressing the <kbd>Up Arrow</kbd> does __NOT__ jump up to the `minValue`. The behavior is different when `minTime < maxTime`; in that case, it is not possible to loop from `minTime` to `maxTime` or vice versa. (In the GIF below, the first couple seconds are me pressing the <kbd>Up Arrow</kbd>.)
min-max-time-bug-3
    - _Expected:_ Do __NOT__ allow looping across the boundaries in either direction.

1. __Setting and resetting `minTime` and `maxTime` in the example:__
    - _Currently:_ The only way to "clear" `minTime` and `maxTime` in the example is to set them to the same value (e.g. set them both to "0:00:00.000"). We also don't really need millisecond precision in the example to demonstrate that `minTime` and `maxTime` work; hour granularity will get the point across.
    - _Expected:_ Again, I'd take inspiration from `numericInputBasicExample.tsx` and use a `pt-select` menu with the following options:
        - Minimum time:
            - `None`
            - `6pm (18:00)`
        - Maximum time:
            - `None`
            - `6pm (18:00)` // for the `minTime === maxTime` case 
            - `9pm (21:00)` // for the `minTime < maxTime` case
            - `2am (02:00)` // for the `minTime > maxTime` case

**Comment by cmslewis** [comment]
Oh, and we should also add some mention in the docs of the `maxTime < minTime` and `minTime === maxTime` handlings. Here are some sample prop descriptions to address that (note the `backticks` around words like "Date" and "minTime" and "maxTime"):

`maxTime`: `Date`
> The latest time the user can select. The year, month, and day parts of the `Date` object are ignored. While the `maxTime` will be later than the `minTime` in the basic case, it is also allowed to be earlier than the `minTime`. This is useful, for example, to express a time range that extends before and after midnight. If the `maxTime` and `minTime` are equal, then the valid time range is constrained to only that one value.

`minTime`: `Date`
> The earliest time the user can select. The year, month, and day parts of the `Date` object are ignored. While the `minTime` will be earlier than the `maxTime` in the basic case, it is also allowed to be later than the `maxTime`. This is useful, for example, to express a time range that extends before and after midnight. If the `maxTime` and `minTime` are equal, then the valid time range is constrained to only that one value.

**Comment by JacekJagiello** [comment]
@cmslewis Thanks for testing out this feature, I'll start to tackling these issues soon ;)

File: packages/datetime/src/common/dateUtils.ts [hunk 24]
```diff
@@ -110,6 +110,55 @@ export function isMonthInRange(date: Date, dateRange: DateRange) {
...
+        return isTimeGreaterThanMinTime || isTimeSmallerThanMaxTime;
+    }
+
+    return isTimeGreaterThanMinTime && isTimeSmallerThanMaxTime;
+}
+
+export function getTimeInRange(time: Date, minTime: Date, maxTime: Date) {
+    if (timesAreEqual(minTime, maxTime)) {
```
Comment (JacekJagiello) [inline packages/datetime/src/common/dateUtils.ts]: I'm not sure if this check is a responsibility of this function, what do you think?

**Comment by JacekJagiello** [comment]
@cmslewis I've fixed two first issues, I'll do `3` and `4` tomorrow.

Check this out: https://16-95210685-gh.circle-artifacts.com/0/home/ubuntu/blueprint/packages/site-docs/dist/index.html#datetime/timepicker

**Comment by JacekJagiello** [comment]
Not ready yet, but I think I'll finish tomorrow.

**Comment by cmslewis** [comment]
No rush! We're all out today for a national holiday anyway.

**Comment by JacekJagiello** [comment]
Ok, I've fixed these issues

Here is build preview: https://21-95210685-gh.circle-artifacts.com/0/home/ubuntu/blueprint/packages/site-docs/dist/index.html#datetime/timepicker

@cmslewis Please review test feature and review code once again  ;)

**Comment by adidahiya** [comment]
nice work @JacekJagiello!
